src/Module.Server/Rewards/CrpgRewardServer.cs

```diff
@@ -195,17 +210,78 @@ public override void OnRemoveBehavior()
             return;
         }
 
+        CalculatePlayerUpkeepCompensations();
+        CompensateTeamhitPlayers(userUpdates);
+
         // TODO: add retry mechanism (the endpoint need to be idempotent though).
         try
         {
-            var res = (await _crpgClient.UpdateUsersAsync(new CrpgGameUsersUpdateRequest { Updates = userUpdates })).Data!;
+            var res = (await _crpgClient.UpdateUsersAsync(new CrpgGameUsersUpdateRequest { Updates = userUpdates.Values.ToList() })).Data!;
             SendRewardToPeers(res.UpdateResults, crpgPeerByUserId, valorousPlayerIds);
         }
         catch (Exception e)
         {
             Debug.Print("Couldn't update users: " + e);
             SendErrorToPeers(crpgPeerByUserId);
         }
+
+        _compensationTable.Clear();
+    }
+
+    private void CalculatePlayerUpkeepCompensations()

```

Comment:
Making this method pure would improve a lot its readability.

Reply:
what do you mean by pure? Just remove the Logs?

Reply:
I meant without side-effects, that is you pass your inputs by argument and return your output.

Reply:
ah so you mean that it should not need any instance variables but rather get them by arguments?
Why is that a bad thing? Sorry for asking i come from a Java background.

Reply:
He means that your function should be able to work in any context. It gets information and output information without touching what it got. so type should not be void.

Reply:
A pure method is way easier to read. You can understand what it's supposed to do just by reading its signature.

Also I don't think there is a good reason to save _compensationTable as a field.

Reply:
Reworked Code and made sure that any Method that can be pure is pure